### ISSUE TITLE
Segfault in threaded pool call to OCILobFreeTemporary

### ISSUE THREAD

**Issue by markfinn** [oracle/python-cx_Oracle#477]
Oracle 12
Linux-3.10.0-1062.9.1.el7.x86_64-x86_64-with-redhat-7.7-Maipo
('sys.maxsize > 2**32:', True)
('platform.python_version:', '2.7.13')
('cx_Oracle.version:', '7.3.0')
('cx_Oracle.clientversion:', (12, 2, 0, 1, 0))"

I have no chance of boiling this down to a reproducible snippet as it is super intermittent and the code is quite involved..  The gist is that I have a thread pool using a Oracle.SessionPool(... threaded=True) to do many writes at a time in separate sessions, and about half of the time those writes involve a blob.

Every so often (say between 100 and 10000 successful writes over the course of 5 minutes to half a day) I get this:

```
ORA-24550: signal received: [si_signo=11] [si_errno=0] [si_code=1] [si_int=-1142876536] [si_ptr=0x7febbbe11688] [si_addr=0x10]

kpedbg_dmp_stack()+400<-
kpeDbgCrash()+210<-
kpeDbgSignalHandler()+121<-
skgesig_sigactionHandler()+272<-
__sighandler()<-
kpuxcRecordOCICall()+105<-
kpuxcRecord_OCILobFreeTemporary()+354<-
kpuxcEntryCallback()+541<-
kpuEntryCallback()+376<-
kpultfr()+9352<-
dpiOci__lobFreeTemporary()+54<-
0000001649DC600C<-
00000000000000A8<-
006C6F6F50697064
```

Is there anywhere I can start debugging?

**Comment by anthony-tuininga** [comment]
Are you actually using temporary LOBs? Those perform a lot poorer than just plain strings/bytes. Can you share the bit of code that you are using to perform your writes? There might be a way to improve that and avoid this issue.

You can also try a 19c client to see if the problem has been resolved in that version. I've seen this sort of thing before but not since I have moved to 19c -- it might be coincidence, though, as I try to avoid the use of temporary LOBs at the best of times!

You can also try being more agreesive in tidying things up sooner. In other words, use `del lob` and `connection.close()` and the like in order to clean things up in the same thread in which they were created/processed. That may (or may not!) help.

Hope some of those suggestions are helpful!

**Comment by markfinn** [comment]
To be honest I don't know if my LOBs are temporary!  I thought there was some reason I couldn't just use a string, but that was a long time ago.  Here's the writes:

```
                            blobvar = c.c.var(cx_Oracle.BLOB, len(data), 1)
                            blobvar.setvalue(0, data)
                            yield c.execute('INSERT INTO xxx (xxx, xxx) '
                                            'VALUES (:in_hash, :in_data)', in_hash=hsh, in_data=blobvar)
```

I do have 19.4.0_64 on those boxes, I just haven't switched to it for production yet. I'll give it a try along with the other suggestions, thanks!

**Comment by anthony-tuininga** [comment]
That code snippet definitely shows that you are using temporary LOBs. You can just do this instead:

```
yield c.execute('INSERT INTO xxx (xxx, xxx) '
                'VALUES (:in_hash, :in_data)', in_hash=hsh, in_data=data)
```

That should perform better and will avoid the problematic OCILobFreeTemporary() call. So that is another option for you to try. :-)

**Comment by markfinn** [comment]
Going to the 19.4.0_64 libraries didn't make it go away.  Getting rid of that particular temporary blobvar did seem to make it go away so far.

I found the reason I originally added that blobvar: when I tried to pass it directly as a string I got "ORA-01465: invalid hex number".  I think originally I fixed it by converting to a blob myself.  This time I tried wrapping the bind variable with  UTL_RAW.CAST_TO_RAW() which seemed to work.

**Comment by anthony-tuininga** [comment]
Ah yes, you can use `cursor.setinputsizes()` to specify that the value is `cx_Oracle.DB_TYPE_LONG_RAW` -- otherwise, it assumes the value is indeed a string. Using `UTL_RAW.CAST_TO_RAW()` should also work but will require some additional effort on the server. In any case, problem solved?

**Comment by markfinn** [comment]
Just for some followup on this, I did still see one segfault after removing the temporary LOBs, but I had gone back to running as a daemon so I didn't capture the stack trace to see if it was still down in the same layer through other free calls.  I've added some stderr logging for the future, but my current high-load dataset is done, so now it might be a while.  I'll close this issue.  Thanks for your help!
